lib/resources/grub_conf.rb

```diff
@@ -71,35 +73,73 @@ def to_s
   ######################################################################
 
   def grub2_parse_kernel_lines(content, conf)
-    # Find all "menuentry" lines and then parse them into arrays
-    menu_entry = 0
+    menu_entries = extract_menu_entries(content)
+
+    if @kernel == 'default'
+      default_menu_entry(menu_entries, conf['GRUB_DEFAULT'])
+    else
+      menu_entries.select { |entry| entry['name'] == @kernel }[0]
+    end
+  end
+
+  def extract_menu_entries(content)
+    menu_entries = []
+
     lines = content.split("\n")
-    kernel_opts = {}
-    kernel_opts['insmod'] = []
-    lines.each_with_index do |file_line, index|
-      next unless file_line =~ /(^|\s)menuentry\s.*/
-      lines.drop(index+1).each do |kernel_line|
-        next if kernel_line =~ /(^|\s)(menu|}).*/
-        if menu_entry == conf['GRUB_DEFAULT'].to_i && @kernel == 'default'
-          if kernel_line =~ /(^|\s)initrd.*/
-            kernel_opts['initrd'] = kernel_line.split(' ')[1]
-          end
-          if kernel_line =~ /(^|\s)linux.*/
-            kernel_opts['kernel'] = kernel_line.split
-          end
-          if kernel_line =~ /(^|\s)set root=.*/
-            kernel_opts['root'] = kernel_line.split('=')[1].tr('\'', '')
-          end
-          if kernel_line =~ /(^|\s)insmod.*/
-            kernel_opts['insmod'].push(kernel_line.split(' ')[1])
-          end
-        else
-          menu_entry += 1
-          break
+    lines.each_with_index do |line, index|
+      next unless line =~ /^menuentry\s+.*/
+      entry = {}
+      entry['insmod'] = []
+
+      # Extract name from menuentry line
+      capture_data = line.match(/(?:^|\s+).*menuentry\s*['|"](.*)['|"]\s*--/)
+      if capture_data.nil? || capture_data.captures[0].nil?
+        raise Inspec::Exceptions::ResourceFailed "Failed to extract menuentry name from #{line}"
+      end
+
+      entry['name'] = capture_data.captures[0]
+
+      # Begin processing from index forward until a `}` line is met
+      lines.drop(index+1).each do |mline|
+        break if mline =~ /^\s*}\s*$/
+        case mline
+        when /(?:^|\s*)initrd.*/
+          entry['initrd'] = mline.split(' ')[1]
+        when /(?:^|\s*)linux.*/
+          entry['kernel'] = mline.split
+        when /(?:^|\s*)set root=.*/
+          entry['root'] = mline.split('=')[1].tr('\'', '')
+        when /(?:^|\s*)insmod.*/
+          entry['insmod'] << mline.split(' ')[1]
         end
       end
+
+      menu_entries << entry
+    end
+
+    menu_entries
+  end
+
+  def default_menu_entry(menu_entries, default)
+    if default == 'saved'
```

Comment:
This is one of those places where I'd probably do this:

```ruby
return menu_entries[default.to_i] unless default == 'saved'

grubenv_contents = ....
```

Anytime I see a single-line else or a really small else block, I'd try and short-circuit the method early and eliminate one level of indentation.

Not a blocker for this, just something to look out for in the future.

Reply:
Great catch! I changed it because I believe that way to be much cleaner.